### PR TITLE
Triaging runtime crash via LLM

### PR DESCRIPTION
**Pull request by fdt622**
# Triaging runtime crash via LLM mentioned in [#221](https://github.com/google/oss-fuzz-gen/issues/221)

To reduce the workload of manually triaging runtime crash, we leverage LLM to identify the cause of crash. We pass fuzz target code and crash information to LLM, ask LLM with designed prompts,and save the reponse from LLM. The crash information contains stack trace and sanitizer output. We can also pass related project code to LLM in the future.

## Conditions to LLM triaging

Currently, triaging runtime crash with LLM would be activated only when all the following conditions are met:

- Build successfully.
- Trigger crash.

## Input to LLM triaging

To retrieve crash information from fuzz log, we add function `extract_crash_info` to class `SemanticCheckResult` in `experiment/fuzz_target_error.py` file. The retrieved information is stored in variable `crash_info` in class `RunResult`. The fuzz target code can be obtained from `result/output-ProjectName-FunctionName/fixed_targets` folder.

## Prompt design

We require that the LLM: 1) definitively ascertain whether the crash is due to errors within the fuzz target or results from a vulnerability in the project under test; 2) deliver a thorough analysis of the findings. Two triage prompts, `triage_priming.txt` and `triage_problem.txt`, are appended to `prompts/template_xml` folder. Three functions, `build_triage_prompt`, `_format_triage_priming`, and `_format_triage_problem`, are added to `llm_toolkit/prompt_builder.py` file. The generated triage prompt would be saved in `result/output-ProjectName-FunctionName/fixed_targets/TargetName-triage` folder.

Below is a generated prompt example:

```plaintext
[{"role": "system", "content": "Given the following crash report and fuzz target, investigate the cause of the crash.\n"}, {"role": "user", "content": "Below is the crash report:\n<log>\n=================================================================\r\n\u001b[1m\u001b[31m==13==ERROR: AddressSanitizer: SEGV on unknown address 0x0000000003a7 (pc 0x000000572ba3 bp 0x7ffd25f0ba10 sp 0x7ffd25f0b970 T0)\r\n\u001b[1m\u001b[0m==13==The signal is caused by a READ memory access.\r\n==13==Hint: address points to the zero page.\r\nSCARINESS: 10 (null-deref)\r\n    #0 0x572ba3 in finalizeTable /src/liblouis/liblouis/compileTranslationTable.c:4377:13\r\n    #1 0x570e1c in _lou_getTranslationTable /src/liblouis/liblouis/compileTranslationTable.c:5059:8\r\n    #2 0x56c574 in LLVMFuzzerTestOneInput /src/liblouis/tests/fuzzing/table_fuzzer.cc:11:50\r\n    #3 0x43de33 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15\r\n    #4 0x43d61a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:514:3\r\n    #5 0x43ece9 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:757:19\r\n    #6 0x43f9b5 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, std::__Fuzzer::allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:895:5\r\n    #7 0x42ed1f in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:912:6\r\n    #8 0x458372 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10\r\n    #9 0x7f72da84a082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 87b331c034a6458c64ce09c03939e947212e18ce)\r\n    #10 0x41f75d in _start (/out/table_fuzzer+0x41f75d)\r\n\r\nDEDUP_TOKEN: finalizeTable--_lou_getTranslationTable--LLVMFuzzerTestOneInput\r\nAddressSanitizer can not provide additional info.\r\nSUMMARY: AddressSanitizer: SEGV /src/liblouis/liblouis/compileTranslationTable.c:4377:13 in finalizeTable\r\n==13==ABORTING\r\nMS: 3 CMP-ChangeByte-InsertRepeatedBytes- DE: \"\\000\\000\\000\\001\"-; base unit: 9fe132914f92e1ad968c7b50f9c606d3922e86c6\r\n0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x23,0x20,0x5d,0x6b,0x6f,0x69,0x0,0x0,0x0,0x1,0x0,0x0,0x0,0x6f,0x4c,0x65,\r\n\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000\\000# ]koi\\000\\000\\000\\001\\000\\000\\000oLe\r\nartifact_prefix='./'; Test unit written to ./crash-578f79fcf8db7fb291a77fb9c20abf8c01eff949\r\nBase64: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACMgXWtvaQAAAAEAAABvTGU=\r\nstat::number_of_executed_units: 938\r\nstat::average_exec_per_sec:     0\r\nstat::new_units_added:          57\r\nstat::slowest_unit_time_sec:    0\r\nstat::peak_rss_mb:              46\r\n\n</log>\n\nBelow is the fuzz target:\n<code>\n#include <fuzzer/FuzzedDataProvider.h>\n#include <liblouis/internal.h>\n\nextern \"C\" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {\n    FuzzedDataProvider stream(data, size);\n\n    // Consume a random length string\n    std::string table_name = stream.ConsumeRandomLengthString();\n\n    // Call the _lou_getTranslationTable function\n    const TranslationTableHeader *table_header = _lou_getTranslationTable(table_name.c_str());\n\n    // Perform some operations with the table_header if needed\n    // For example, checking if the table_header is not NULL\n    if (table_header != nullptr) {\n        // Do something with the table_header\n    }\n\n    // Return 0 to indicate successful execution\n    return 0;\n}\n</code>\n\nTriage the crash:\nFirstly, you must clearly determine whether the crash is caused by errors within the fuzz target, such as incorrect function usage, improper parameter initializations, missing setup steps, etc., or if it stems from a vulnerability in the project under test.\nSecondly, you are required to provide a comprehensive analysis of your findings.\n<solution>"}]
```

## LLM triaging

One function, `triage_crash`, is added to `experiment/evaluator.py` file. The core component of LLM triage, `crash_triage.py`, is appended to `llm_toolkit` folder. `crash_triage.py` mainly contains functions `llm_triage` and `apply_llm_triage`.

## Response from LLM triaging

One function, `parse_triage_response`, is appended to `llm_toolkit/output_parser.py` file. The generated triage response from LLM would be saved in `result/output-ProjectName-FunctionName/fixed_targets/TargetName-triage` folder.

Below is a reponse example:

```plaintext
### Triage Analysis:

1. **Crash Cause:** The crash occurred in the `finalizeTable` function at line 4377 of `compileTranslationTable.c`. The crash is a segmentation fault (SEGV) caused by a READ memory access at an unknown address.

2. **Fuzz Target Analysis:**
   - The fuzz target reads input data using `FuzzedDataProvider` and extracts a random length string as `table_name`.
   - It then calls the `_lou_getTranslationTable` function with `table_name` as input.
   - The crash happens during the execution of the fuzz target, specifically within the `_lou_getTranslationTable` function.

3. **Root Cause Analysis:**
   - The crash is likely due to a bug in the `_lou_getTranslationTable` function or its interaction with the provided `table_name`.
   - The fuzz target does not seem to have any obvious issues that could directly cause a crash, such as invalid memory accesses or incorrect function usage.

4. **Next Steps:**
   - To investigate further, you need to analyze the implementation of the `_lou_getTranslationTable` function in the `compileTranslationTable.c` file.
   - Look for potential issues such as NULL pointer dereferences, out-of-bounds memory accesses, or incorrect handling of input data.
   - Consider running the fuzz target with different inputs to see if the crash is consistently reproducible.
   - Use debugging tools like AddressSanitizer, Valgrind, or GDB to pinpoint the exact cause of the crash within the `_lou_getTranslationTable` function.

5. **Recommendations:**
   - Review the implementation of `_lou_getTranslationTable` function for any potential vulnerabilities.
   - Validate input data handling within the function to ensure it can handle all possible inputs gracefully.
   - Test the function with various inputs, including edge cases, to uncover any hidden issues.
   - Consider adding additional checks and error handling mechanisms to prevent crashes in case of unexpected input scenarios.

6. **Conclusion:**
   - The crash is most likely caused by an issue within the `_lou_getTranslationTable` function or its interaction with the provided input data.
   - Further investigation into the function's implementation and input handling is necessary to identify and fix the root cause of the crash.
```

## Additions within `results` folder

`result/output-ProjectName-FunctionName/fixed_targets/TargetName-triage` folder is created to store the generated prompt file and trirage rawoutput from LLM.

## Additions within results report

Triage report is appended to results report web. We add a column `Triage` under `Run logs` to store triage reports. One function, `get_triage`, is added to `report/web.py` file.

### CODE REVIEW

**Review by fdt622 — COMMENTED** [review]
checked.

File: prompts/template_xml/triage_priming.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+Given the following crash report and fuzz driver, investigate the cause of the crash.
```
Comment (DonggeLiu) [inline prompts/template_xml/triage_priming.txt:1]: Do we need to ask LLM to keep the response succinct?
Sometimes, I found it returned a wall of text with a lot of useless info.

File: prompts/template_xml/triage_problem.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+{CRASH_REPORT}
+</log>
+
+Below is the fuzz driver:
+<code>
+{FUZZ_TARGET_CODE}
+</code>
+
+Based on the above information, analyze the cause of the crash. First, only answer “the crash is caused by the bug in the fuzz driver” or “the crash is caused by the bug in the project”. Second, give analysis.
+<solution>
```
Comment (DonggeLiu) [inline prompts/template_xml/triage_problem.txt:12]: I reckon lines 11-12 are general requirements, hence belong to priming?
This `problem` should be a single problem.

Imagine we will have some example triage problem-solution pairs later: This template will be used to formulate the sample problem, and we don't need to repeat this in each sample problem.

File: report/templates/sample.html [hunk 3]
```diff
@@ -49,4 +49,9 @@ <h2>Run logs</h2>
 <pre>
 {{ run_logs }}
 </pre>
+
+<h2>Triage</h2>
+<pre>
+{{ triage }}
+</pre>
```
Comment (DonggeLiu) [inline report/templates/sample.html:56]: Would it be a good idea to move this to the top of the file?
I reckon this is a pretty important conclusion about the fuzz target to state at the beginning, what do you think?

File: prompts/template_xml/triage_problem.txt (hunk 2 above)
Comment (DonggeLiu) [reply prompts/template_xml/triage_problem.txt:12]: nit: s/fuzz driver/fuzz target/g

File: report/templates/sample.html (hunk 3 above)
Comment (DonggeLiu) [reply report/templates/sample.html:56]: BTW, shall we also add a new column at [benchmark.html](https://github.com/google/oss-fuzz-gen/blob/main/report/templates/benchmark.html#L33) to show if LLM thinks this crash is caused by the fuzz target or the project?

File: llm_toolkit/crash_triage.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+def triage_all_crashes():
+  """Reads crash information and relevant code, applies triage, \
+        and saves the result."""
+  # TODO(fdt622): Finish this.
+
+
+# ========================= LLM Triage ========================= #
+
+
+def llm_triage(
```
Comment (DonggeLiu) [inline llm_toolkit/crash_triage.py:83]: Thanks for following the same structure as `llm_fix`, this simplifies readability and maintainability.
Also, we might need to convert this to `llm_fix` later, if we want to test asking LLM to fix code based on its own triage response : )

**Review by DonggeLiu — COMMENTED** [review]
Thanks for the code contribution and detailed explanation, @fdt622!
They look good to me overall, I left some minor changes suggestions and nit comments, hope you won't mind : )

Please let me know if my comments make sense or if I missed anything, happy to chat further about them.
Thanks!

**Comment by DonggeLiu** [comment]
BTW could you please also resolve the conflict if you have time?
Sorry that we pushed some changes before reviewing your code.

File: experiment/fuzz_target_error.py [hunk 5]
```diff
@@ -44,6 +44,9 @@ class SemanticCheckResult:
   # E.g., matches 'SCARINESS: 10 (null-deref)'
   SYMPTOM_SCARINESS = re.compile(r'SCARINESS:\s*\d+\s*\((.*)\)\n')
 
+  # Regex for extract crash information.
+  INFO_CRASH = re.compile(r'^={65}\r?\n([\s\S]*)', re.MULTILINE | re.DOTALL)
+
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/fuzz_target_error.py:49]: Your concern is not unreasonable. Based on my observations so far, the 65 consecutive equals sign is applicable. I have observed some projects that have more than 65 consecutive equal signs interfering in their run results logs, so I'll use 65 equal signs as a dividing line for now.

File: .gitignore [hunk 6]
```diff
@@ -8,3 +8,4 @@ venv/
 .venv/
 __pycache__/
 *.swp
+.DS_Store
```
[the remark this answers is not in the document]
Comment (fdt622) [reply .gitignore:11]: Thank you for your careful review, I will modify it one by one.

This one done.

File: experiment/builder_runner.py [hunk 7]
```diff
@@ -204,7 +205,7 @@ def _stack_func_is_of_testing_project(self, stack_frame: str) -> bool:
             LIBFUZZER_LOG_STACK_FRAME_CPP not in stack_frame)
 
   def _parse_libfuzzer_logs(
-      self, log_handle) -> tuple[int, int, bool, SemanticCheckResult]:
+      self, log_handle) -> tuple[int, int, bool, str, SemanticCheckResult]:
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/builder_runner.py:208]: Done.

File: experiment/builder_runner.py [hunk 8]
```diff
@@ -239,38 +241,41 @@ def _parse_libfuzzer_logs(
     initcov, donecov, lastround = self._parse_fuzz_cov_info_from_libfuzzer_logs(
         lines)
 
+    crash_info = ''
+
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/builder_runner.py:245]: Done.

File: experiment/evaluator.py [hunk 9]
```diff
@@ -218,6 +218,27 @@ def _fix_generated_fuzz_target(self, ai_binary: str,
...
+  def triage_crash(
+      self,
+      ai_binary: str,
+      generated_oss_fuzz_project: str,
+      triaged_target_path: str,
+      run_result: RunResult,
+      logger: _Logger,
+  ):
+    """Triage the crash."""
+    if run_result.crash_info == '':
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/evaluator.py:230]: Done.

File: experiment/evaluator.py [hunk 10]
```diff
@@ -218,6 +218,27 @@ def _fix_generated_fuzz_target(self, ai_binary: str,
...
+      logger.log(f"Warning: no crash info in {generated_oss_fuzz_project}.")
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/evaluator.py:231]: Done.

File: experiment/evaluator.py [hunk 11]
```diff
@@ -218,6 +218,27 @@ def _fix_generated_fuzz_target(self, ai_binary: str,
...
 
+  def triage_crash(
+      self,
+      ai_binary: str,
+      generated_oss_fuzz_project: str,
+      triaged_target_path: str,
+      run_result: RunResult,
+      logger: _Logger,
+  ):
+    """Triage the crash."""
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/evaluator.py:229]: Done.

File: llm_toolkit/crash_triage.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:1]: Done.

File: llm_toolkit/crash_triage.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+  os.makedirs(args.intermediate_output_dir, exist_ok=True)
+
+  return args
+
+
+def triage_all_crashes():
+  """Reads crash information and relevant code, applies triage, \
+        and saves the result."""
+  # TODO(fdt622): Finish this.
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:77]: Done.

File: llm_toolkit/crash_triage.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+
+
+# ========================= LLM Triage ========================= #
+
+
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:82]: Done.

File: llm_toolkit/crash_triage.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  triage_candidates = []
+  for file in os.listdir(response_dir):
+    if not parser.is_raw_output(file):
+      continue
+    triage_path = os.path.join(response_dir, file)
+    triage = parser.parse_triage(triage_path)
+    triage_candidates.append([triage_path, triage])
+
+  if not triage_candidates:
+    print(f'LLM did not generate rawoutput for {prompt_path}.')
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:116]: Done.

File: llm_toolkit/crash_triage.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+    return
+
+  # TODO(fdt622): Use the common vote
+  # Currently, we prefer the longest triage.
+  preferred_triage_path, preferred_triage = max(triage_candidates,
+                                                key=lambda x: len(x[1]))
+  print(
+      f'Will use the longest triage: {os.path.relpath(preferred_triage_path)}.')
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:124]: Done.

File: llm_toolkit/output_parser.py [hunk 17]
```diff
@@ -84,6 +84,17 @@ def parse_code(response_path: str) -> str:
...
 
+def parse_triage(triage_path: str) -> str:
+  """Parses the triage from the |triage_path|."""
+  with open(triage_path) as file:
+    triage = file.read()
+  solution = triage.split("</solution>")[0]
+
+  lines = solution.splitlines()
+
+  return "\n".join(lines)
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/output_parser.py:95]: Done.

File: llm_toolkit/prompt_builder.py [hunk 18]
```diff
@@ -329,6 +333,35 @@ def _format_fixer_problem(self, raw_code: str, error_desc: Optional[str],
...
+                                          priming_weight)
+
+    self._prepare_prompt(priming, problem)
+    return self._prompt
+
+  def _format_triage_priming(self) -> Tuple[str, int]:
+    """Formats a priming for crash triage based on the template."""
+    with open(self.triage_priming_template_file) as f:
+      priming = f.read().strip() + "\n"
+    priming_prompt = self._prompt.create_prompt_piece(priming, "system")
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/prompt_builder.py:350]: Done.

File: llm_toolkit/crash_triage.py (hunk 4 above)
Comment (fdt622) [reply llm_toolkit/crash_triage.py:83]: Okay, I'll convert it when we need to. For now I keep it as llm_triage.

File: prompts/template_xml/triage_problem.txt (hunk 2 above)
Comment (fdt622) [reply prompts/template_xml/triage_problem.txt:12]: Sorry, I didn't fully understand what you meant, especially "nit: s/fuzz driver/fuzz target/g". The following are modifications based on my understanding:

priming:
https://github.com/fdt622/oss-fuzz-gen/raw/dev/prompts/template_xml/triager_priming.txt

problem:
https://github.com/fdt622/oss-fuzz-gen/raw/dev/prompts/template_xml/triager_problem.txt

If other modifications are needed or I have misunderstood something, please point it out.

File: prompts/template_xml/triage_priming.txt (hunk 1 above)
Comment (fdt622) [reply prompts/template_xml/triage_priming.txt:1]: Yes, the problem you mentioned does exist. I tried to make the following modifications and asked LLM to give a concise answer. But at the same time, after the modification, in some cases LLM will answer too concisely. We may need better prompts in the future.

https://github.com/fdt622/oss-fuzz-gen/blob/0de21f16d9254cc7c4fff782276d48741bab9643/prompts/template_xml/triager_priming.txt#L3

File: llm_toolkit/prompt_builder.py [hunk 19]
```diff
@@ -329,6 +333,35 @@ def _format_fixer_problem(self, raw_code: str, error_desc: Optional[str],
...
 
+  def build_triage_prompt(self, benchmark: Benchmark, target_code: str,
+                          crash_info: str) -> prompts.Prompt:
+    """Prepares the crash-triaging prompt."""
+    priming, priming_weight = self._format_triage_priming()
+    problem = self._format_triage_problem(target_code, crash_info,
+                                          priming_weight)
+
+    self._prepare_prompt(priming, problem)
+    return self._prompt
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/prompt_builder.py:344]: Not yet. The good triage example I have seen is that after I manually provide the project source code of the relevant stack, I will try to find a good triage example after providing the project source code of the relevant crash stack to LLM.

File: llm_toolkit/crash_triage.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+  # TODO(fdt622): go through and modify build_triage_prompt
+  prompt = builder.build_triage_prompt(benchmark, target_code, crash_info)
+  prompt.save(prompt_path)
+
+  triage_model.generate_code(prompt, response_dir)
+
+
+def main():
+  #TODO(fdt622): Finish this.
+  return 0
```
[the remark this answers is not in the document]
Comment (fdt622) [reply llm_toolkit/crash_triage.py:159]: Thank you for your comments above regarding the crash_triager.py file, I apologize for many vacancies in that file. I would like to temporarily disable the use of that file as a separate script.

File: experiment/fuzz_target_error.py [hunk 21]
```diff
@@ -69,6 +72,15 @@ def is_no_cov_increase_err(cls, error_desc: Optional[str]) -> bool:
...
         cls.NO_COV_INCREASE_MSG_PREFIX)
 
+  @classmethod
+  def extract_crash_info(cls, fuzzlog: str) -> str:
+    """Extracts crash information from fuzzing logs."""
+    match = cls.INFO_CRASH.search(fuzzlog)
+    if match:
+      return match.group(0)
+
+    return ''
```
[the remark this answers is not in the document]
Comment (fdt622) [reply experiment/fuzz_target_error.py:82]: Thanks for your advice. To not make parameter passing more complicated, I added a simple warning without mentioning the object name for now.

**Comment by fdt622** [comment]
> BTW could you please also resolve the conflict if you have time? Sorry that we pushed some changes before reviewing your code.

Hi, I have modified and merged the code. There are two main changes. 

One is to add a triage column to benchmark.html:
Mainly, I create class ["TriageResult"](https://github.com/fdt622/oss-fuzz-gen/blob/ade92be31ad1d106f6b7dcb4f1dbb1f73a7afec9/llm_toolkit/crash_triager.py#L26), add variable "triage" in [evaluator.py](https://github.com/fdt622/oss-fuzz-gen/blob/ade92be31ad1d106f6b7dcb4f1dbb1f73a7afec9/experiment/evaluator.py#L51) and [builder_runner.py](https://github.com/fdt622/oss-fuzz-gen/blob/ade92be31ad1d106f6b7dcb4f1dbb1f73a7afec9/experiment/builder_runner.py#L94).

The other is to add return statement at the final part of "if crash" condition:
https://github.com/fdt622/oss-fuzz-gen/blob/ade92be31ad1d106f6b7dcb4f1dbb1f73a7afec9/experiment/builder_runner.py#L332
This is to make sure ”crash_info“ is returned for further triage as long as the crash is triggered.

Please tell me if further changes are needed. Thank you.

**Comment by DonggeLiu** [comment]
Hi @fdt622 it appears our new PR-exp CI can work directly on this PR, I've started one.
While waiting, could you please fix the linter error?
You can also check it locally via:
https://github.com/google/oss-fuzz-gen/blob/main/USAGE.md#development-environment

**Comment by fdt622** [comment]
Hi @DonggeLiu , I have fixed the linter error.

> Hi @fdt622 ~it appears our new PR-exp CI can work directly on this PR, I've started one.~ It cannot pull the image so I guess we have to use the old trick. While waiting, could you please fix the linter error? You can also check it locally via: https://github.com/google/oss-fuzz-gen/blob/main/USAGE.md#development-environment

**Comment by fdt622** [comment]
Hi @DonggeLiu , sorry for the linter error. After fixing no-else-return, there are still two unused-arguments in my code (benchmark and priming_weight). I think maybe we can keep them, because I will use them in the next PR that add source code to prompt. As for the other unused-argument (check_cov_increase), it belongs to someone else.

**Comment by DonggeLiu** [comment]
Thanks, @fdt622!
